### ISSUE TITLE
IV was not an issue

### ISSUE THREAD

**Issue by aehrath-gretel** [capitalone/fpe#23]
I dug deeper and apparently there is no way to get the IV back unless it is stored since the order is reversed.

**Comment by anitgandhi** [comment]
Further, the IV is defined in the FF1 spec to be hardcoded to 0. So we couldn't actually let the user decide what it would be.

**Comment by anitgandhi** [comment]
Well, it's not really regular ECB. `ciph()` implements CBC, ultimately, but it can be called with 1 or more blocks of input. 

CBC with a single block and a null (zero) IV is algorithmically equivalent to ECB, because there's no subsequent block to chain into.

CBC with multiple blocks and a null (zero) IV is NOT algorithmically equivalent to ECB, because the output of the first encryption is chained into the next block. Whereas regular ECB on multiple blocks just encrypts each block separately, with no connection between each block. 

So, if you're porting this to Python, ensure you're using CBC with IV=0, not ECB, and you should see the NIST test vectors pass. 

> Do you know if ECB is not as safe as CBC with the IV set to 0 all the time?

Under normal circumstances, yes ECB is not considered secure. But ECB/CBC here is used as a primitive or a building block to construct a secure algorithm, as designed by the authors and reviewed.

**Comment by aehrath-gretel** [comment]
"CBC with a single block and a null (zero) IV is algorithmically equivalent to ECB, because there's no subsequent block to chain into."

Right. What I meant was: Since CBC with IV Zero every time is basically ECB, but ECB runs a lot faster since you do not need to run the XOR, or the IV init every time, why not use ECB in the first place and add another set to the NIST standard using ECB?

**Comment by anitgandhi** [comment]
Because it's not always a single block that is coming in. The number of blocks of input is an important factor for the behavior of CBC. 

Looking through the code, there are 3 places where `ciph()` is called:

1. [here](https://github.com/capitalone/fpe/blob/e6bfa0dae121111c2e93483777f0f2cde562f341/ff1/ff1.go#L324), on a single block input
2. [here](https://github.com/capitalone/fpe/blob/e6bfa0dae121111c2e93483777f0f2cde562f341/ff1/ff1.go#L558), on a single block input
3. [here](https://github.com/capitalone/fpe/blob/e6bfa0dae121111c2e93483777f0f2cde562f341/ff1/ff1.go#L613), on a variable/multi block input. 

For (1) and (2), yes, you could safely shortcut this by directly doing ECB. For (3), it requires CBC because it's multiple blocks, and the spec defines `prf()`` as essentially CBC-MAC. If you replaced this with ECB, it wouldn't implement what the spec defines. 

There are 2 reasons why I did not do so for (1) and (2):

1. The code is easier to follow in the current way such that the function names more closely match what's defined in the spec, making it easier to follow. 

2. Because there's only 1 block, the Go stdlib CBC implementation (or any CBC implementation, for that matter) has nothing to do after the first block. While I haven't tested this directly, I doubt that it would actually be faster to use ECB rather than CBC for the single block scenario. 

> why not use ECB in the first place and add another set to the NIST standard using ECB?

As to why the algorithm authors decided ECB was not sufficient here, I can't say with full confidence, but here's what the spec says, in section 4.6:

```
The PRF(X) function, specified in Algorithm 6, essentially invokes the Cipher Block Chaining encryption mode [11] on the input bit string and returns the final block of the ciphertext; this function is the pseudorandom core of the Feistel round function for FF1.Encrypt and FF1.Decrypt
```

Extrapolating from that, I imagine that ECB doesn't provide a strong enough pseudorandom core for the Feistel round.

**Comment by aehrath-gretel** [comment]
First of all, big thank you for your responsiveness!

So I ported all of the FF1 code to Python for our use case. I noted, that you set the IV to 0 [here.](https://github.com/capitalone/fpe/blob/e6bfa0dae121111c2e93483777f0f2cde562f341/ff1/ff1.go#L605)
From what I can tell stepping through the code and looking up where it is being called from, it only ever gets called with 32bytes at the most. This means it only runs with 2 blocks at most. It feels that there is not much security gained in this specific use case.

Another thing I noticed, I was able to set the IV to Zero only once, before you start the FF1 process, so that each block in the Feistel rotation had a completely different value except for the first block of course which has Zero. It still decompresses fine if you do the same thing and just set the first block to IV 0 before you run the Feistel and then leave it alone. You might want to try and comment out that line and see if it still works. It will of course not produce the same values as the NIST, but it will encrypt and decrypt properly, probably with more security since you have more unique IVs. It also runs faster.

To summarize, I do not think in its current form more than 2 blocks get encrypted before the IV gets set to 0, and the SetIV to Zero needs only be done once before the Feistel gets started.

I really appreciate your thoughts on this.

-Alex

**Comment by anitgandhi** [comment]
> It will of course not produce the same values as the NIST, but it will encrypt and decrypt properly, probably with more security since you have more unique IVs. It also runs faster.

Keep in mind that the correctness of test vectors is vitally important. It's not sufficient for an algorithm implementation (cryptographic or otherwise) to simply be correct when doing the inverse (encrypt then decrypt matching the original plaintext). 

If the encrypted values don't match, the implementation is not compliant, which makes interchanging values across implementations impossible. Again, note that this isn't anything specific to FPE, the same can be said of really any algorithm. 

Imagine, if you will, good old AES - if one implementation was "optimized" in a way that encryption then decryption works to get the original value back. But the ciphertext is incorrect - if you took that ciphertext and passed it into a different (compliant) decryption implementation, it would fail. Hence, it's a MUST that all values match the NIST test vectors - it's what determines compliance of an implementation to the spec. 

I can understand where you're coming from w.r.t different IVs being related to higher security when you look at other algorithms (AES CBC, GCM [IV being nonce in this case], etc) but the important difference is that those algorithms explicitly define their IV as being variable and explain the consequences of static IVs. In the context of FF1/FF3, the algorithm explicitly defines the IV to be 0, because it's really just [CBC-MAC](https://en.wikipedia.org/wiki/CBC-MAC). 

All that said, I will look into moving the `setIV(0)` call out of `ciph()` to reduce the number of calls, but considering what that function does is just a `copy`, I don't expect the performance to improve by a meaningful amount. It definitely won't reduce the number of allocations, which tends to be one of the biggest drivers of performance degradation.
